Fix Citybikes naming

aronsky opened the pull request:
## Description:

The recent changes in Home Assistant core broke the naming scheme for CityBikes sensors. This PR fixes the naming.

Basically, up until now, Citybikes assigned each bike-sharing station a name based on the network and station IDs, to be used by HASS for generation of the entity ID of the station. The friendly name attribute was used to provide a nice, readable alternative to this ID.

However, following some recent changes (I believe PR #12237), the custom friendly name seems to be discarded by HASS, if the name attribute exists.

In over to overcome this, Citybikes now provides an `entity_id` property based on the previous template of network ID and station ID, while the `name` property is used to provide a friendly name, when available.

This change may be breaking, depending on user configuration, as the user-configurable name of the station cluster is not incorporated in the entity ID anymore.


**Related issue (if applicable):**
No issue was created, a direct fix is proposed here instead.

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

aronsky committed:
Fix Citybikes naming.

MartinHjelmare commented:
The problem is that the entity defines a friendly name attribute in the device state attributes. That's not the correct way. The entity should only define the friendly name in the `name` property.

OttoWinter commented:
Basically the same as #12346; however the reason why this is happening is different from what you explained in the description of your PR if I understand correctly. Since the introduction of the entity registry `attributes[ATTR_FRIENDLY_NAME]` is no longer used for the displayed friendly name - instead only the `name` **property** is used (see https://github.com/home-assistant/home-assistant/pull/12346#issuecomment-365027885).

MartinHjelmare commented:
Yes, but you added the better context.

aronsky commented:
I see. No one commented on the use of `ATTR_FRIENDLY_NAME` before, when I first created the sensor, so I assumed that's how it was done. Anyway, with this change, the friendly name is reported via the `name` attribute, so should be in line with the current guidelines.

OttoWinter reviewed (commented):
You're still setting `attributes[ATTR_FRIENDLY_NAME]` [here](https://github.com/home-assistant/home-assistant/pull/12661/files#diff-6c32aba97e8731a561a6d4a99d80b603R287), I think you should remove that to remove potential confusion.

aronsky commented on homeassistant/components/sensor/citybikes.py:
```diff
@@ -250,13 +251,19 @@ def state(self):
         return self._station_data.get(ATTR_FREE_BIKES, STATE_UNKNOWN)
 
+    @property
+    def entity_id(self):
```
Thanks for the input! My initial idea was to use `unique_id` (which seems to be the perfect field to store this kind of information). However, since the changes in the way entity IDs are applied, the code in Home Assistant prefers to use the `name` property for entity ID generation, as opposed to `unique_id` (i.e., if the `name` property isn't `None`, it will use it to generate `entity_id`, even if `unique_id` exists). See here: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity_platform.py#L197

I feel that for this platform (it shows the number of bikes available at different bike-sharing stations), an entity ID built from the bike sharing network ID/name and the network-specific station ID is more appropriate than the name of the station (which is usually an address). Maybe I'm wrong and not up to date on the philosophy behind entity IDs. However, the address is definitely more appropriate as a friendly name.

As for your concern about duplicate entity IDs, this shouldn't be a problem, as the `entity_id` property is rewritten by HASS anyway, as you can see here: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity_platform.py#L222

aronsky commented:
I'm open to changing the platform to use the station names as entity IDs.

The original idea was to generate the entity ID based on the bike sharing network name and the station ID while using the station name (usually, its address) as the friendly name. If people think that it's appropriate to use the name/address as the basis for the entity ID, I'm fine with that. I do insist on using the name/address as the friendly name that appears in the UI, as it is much more usable.

MartinHjelmare commented on homeassistant/components/sensor/citybikes.py:
```diff
@@ -13,7 +13,8 @@
 import homeassistant.helpers.config_validation as cv
-from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.components.sensor import (
+    PLATFORM_SCHEMA, DOMAIN as PARENT_DOMAIN)
```
Instead of importing as `PARENT_DOMAIN`, change the name of the existing `DOMAIN` constant in this module. You're not allowed to overwrite the component domain in the platform, in this case the sensor domain. Eg:
```py
CITYBIKES_DATA = 'citybikes'
```

aronsky replied:
Fixed that.

MartinHjelmare commented on homeassistant/components/sensor/citybikes.py:
```diff
@@ -250,13 +251,19 @@ def state(self):
+    @property
+    def entity_id(self):
+        """Return the entity ID."""
+        return "{}.{}_{}".format(PARENT_DOMAIN,
```
Remove the property and define an instance attribute `self.entity_id` in init instead. This isn't doing any I/O, correct?

MartinHjelmare replied:
And it is recommended to use `generate_entity_id` as @OttoWinter says.

Only entities that defines a unique ID get registered in the entity registry. So id conflict can still occur for all other entities.

aronsky replied:
For some reason, calling `generate_entity_id` gets stuck. I think it has something to do with the fact that the platform setup is async, but the call to `generate_entity_id` is performed in a constructor, which is not async. From looking at the code of other platforms and components, it seems that only platforms that aren't async use `generate_entity_id`, while others that need the functionality, simply create the ID on their own - which is what I've done here (after all, the logic for creating the ID is pretty straightforward, and shouldn't result in any duplicates).

OttoWinter replied:
Then you can just call `async_generate_entity_id` ;)

aronsky replied:
I tried that, that doesn't work either... :-/...
I'm supposed to `yield from` it, am I not? This is what I tried in the constructor:
`self.entity_id = yield from async_generate_entity_id(ENTITY_ID_FORMAT, uid, hass=hass)`

It results in an exception:
```
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/helpers/entity_platform.py", line 84, in async_setup
    SLOW_SETUP_MAX_WAIT, loop=hass.loop)
  File "/usr/local/lib/python3.6/asyncio/tasks.py", line 358, in wait_for
    return fut.result()
  File "/usr/local/lib/python3.6/asyncio/futures.py", line 245, in result
    raise self._exception
  File "/usr/local/lib/python3.6/asyncio/tasks.py", line 180, in _step
    result = coro.send(None)
  File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/components/sensor/citybikes.py", line 172, in async_setup_platform
    devices.append(CityBikesStation(hass, network, station_id, name))
TypeError: __init__() should return None, not 'generator'
```

And when using the normal version, as I mentioned before, it just hangs inside the execution of `generate_entity_id`...

MartinHjelmare replied:
Yeah, sorry for unclear comments. You can't `yield from` in the constructor. So you have to do generate the entity_id in `async_setup_platform` and pass the id as an argument to the constructor.

aronsky replied:
OK, after more experimentations, tried calling `async_generate_entity_id` directly (without `yield from`), and it worked. I'm confused - I thought I'm supposed to `await` or to `yield from` any coroutines...

MartinHjelmare replied:
It's not a coroutine, it's a callback.

aronsky replied:
It seems that the naming is a bit confusing, in that case... Oh well.

MartinHjelmare replied:
A function that is named `async_*` in home assistant means that the function needs to run in the event loop. Coroutines are marked with the `@asyncio.coroutine` annotation or defined with the python 3.5 keyword `async def`.

aronsky replied:
I see. Thanks!

aronsky committed:
Semantic fixes upon requests.

Entity ID generation now includes the base name (again),
resulting in preservation of the entity ID format (the change is
non-breaking)

aronsky committed:
Use `async_generate_entity_id`.

MartinHjelmare reviewed (approved):
Thanks! Can be merged when build passes.

aronsky commented:
@MartinHjelmare sorry for the confusion. As I mentioned in one of the comments, the change ended up not being breaking after all, so no need to mark it as breaking in release notes.

MartinHjelmare commented:
Great!
